pkg/checks/traceroute/check.go

```diff
@@ -0,0 +1,195 @@
+	// The path taken to the destination
+	Hops map[int][]Hop `json:"hops" yaml:"hops" mapstructure:"hops"`
+}
+
+// Run runs the check in a loop sending results to the provided channel
+func (tr *Traceroute) Run(ctx context.Context, cResult chan checks.ResultDTO) error {
+	ctx, cancel := logger.NewContextWithLogger(ctx)
+	defer cancel()
+	log := logger.FromContext(ctx)
+
+	log.Info("Starting traceroute check", "interval", tr.config.Interval.String())
+	for {
+		select {
+		case <-ctx.Done():
+			log.Error("Context canceled", "error", ctx.Err())
+			return ctx.Err()
+		case <-tr.DoneChan:
+			return nil
+		case <-time.After(tr.config.Interval):
+			res := tr.check(ctx)
+			tr.metrics.MinHops(res)
+			cResult <- checks.ResultDTO{
+				Name: tr.Name(),
+				Result: &checks.Result{
+					Data:      res,
+					Timestamp: time.Now(),
+				},
+			}
+			log.Debug("Successfully finished traceroute check run")
+		}
+	}
+}
+
+// GetConfig returns the current configuration of the check
+func (tr *Traceroute) GetConfig() checks.Runtime {
+	tr.Mu.Lock()
+	defer tr.Mu.Unlock()
+	return &tr.config
+}
+
+func (tr *Traceroute) check(ctx context.Context) map[string]result {
+	res := make(map[string]result)
+	log := logger.FromContext(ctx)
+
+	type internalResult struct {
+		addr string
+		res  result
+	}
+
+	cResult := make(chan internalResult, len(tr.config.Targets))
+	var wg sync.WaitGroup
+
+	wg.Add(len(tr.config.Targets))
+	for _, t := range tr.config.Targets {
+		go func(t Target) {
+			defer wg.Done()
+			l := log.With("target", t.Addr)
+			l.Debug("Running traceroute")
+
+			start := time.Now()
+			trace, err := tr.traceroute(ctx, tracerouteConfig{
+				Dest:    t.Addr,
+				Port:    int(t.Port),
+				Timeout: int(tr.config.Timeout / time.Millisecond),
+				MaxHops: tr.config.MaxHops,
+				Rc:      tr.config.Retry,
+			})
+			elapsed := time.Since(start)
+			if err != nil {
+				l.Error("Error running traceroute", "error", err)
+			}
+
+			tr.metrics.CheckDuration(t.Addr, elapsed)
+
+			l.Debug("Ran traceroute", "result", trace, "duration", elapsed)
+			res := result{
+				Hops:    trace,
+				MinHops: tr.config.MaxHops,
+			}
+
+			for ttl, hop := range trace {
+				for _, attempt := range hop {
+					if attempt.Reached && attempt.Ttl < res.MinHops {
+						res.MinHops = ttl
+					}
+				}
+			}
+
+			cResult <- internalResult{addr: t.Addr, res: res}
+		}(t)
+	}
+
+	wg.Wait()
+	close(cResult)
+
+	for r := range cResult {
+		res[r.addr] = r.res
+	}
+
+	return res
+}
+
+// Shutdown is called once when the check is unregistered or sparrow shuts down
+func (tr *Traceroute) Shutdown() {
+	tr.DoneChan <- struct{}{}
+	close(tr.DoneChan)
+}
+
+// SetConfig is called once when the check is registered
+// This is also called while the check is running, if the remote config is updated
+// This should return an error if the config is invalid
+func (tr *Traceroute) SetConfig(cfg checks.Runtime) error {
```

Comment:
Maybe we should log a warning if the binary has no capabilities set to signalize the user as soon as he wants to use the check that he may have forgot to set the capabilities. To check for capabilities, you can refer to my [implementation](https://github.com/caas-team/sparrow/blob/feat/tcp-traceroute/internal/helper/capabilities.go#L16) of the traceroute check:
```go
// Capability represents a Linux capability
type Capability uint32

const (
	// CAP_NET_RAW is the capability to use raw sockets
	CAP_NET_RAW Capability = (1 << uint32(13))
)

// HasCapabilities checks if the current process has the specified capabilities.
func HasCapabilities(ca Capability) bool {
	if unix.Geteuid() == 0 {
		return true
	}

	hdr := &unix.CapUserHeader{
		Pid:     int32(unix.Getpid()),
		Version: unix.LINUX_CAPABILITY_VERSION_3,
	}
	data := &unix.CapUserData{}

	err := unix.Capget(hdr, data)
	if err != nil {
		return false
	}

	return data.Effective&uint32(ca) != 0
}
```

Reply:
I decided against doing a check like this, since I'm not 100% sure if theyre are edge cases where this would return false negative. I don't know if those are the only two ways of checking if the process is allowed CAP_NET_RAW. To me it feels safer to just ask the os directly: "Can I open this socket" instead. That way we don't have to try and mirror the kernels authz logic